syntax/text/parser/src/Luna/Syntax/Text/Parser/Parsing.hs

```diff
@@ -919,20 +921,98 @@ impSrc = buildAsg $ (\s -> liftIRBApp0 $ IR.unresolvedImpSrc' s) <$> (wrd <|> re
     rel = Import.Relative <$  symbol Lexer.Accessor <*> qualConsName
     abs = Import.Absolute <$> qualConsName
 
+------------------------------------
+-- === Foreign Import Parsing === --
+------------------------------------
+
+foreignImportList :: AsgParser SomeExpr
+foreignImportList = buildAsg $
+    (\lang imports ->
+        liftAstApp1 (IR.foreignImpList' lang) (sequence imports))
+    <$  (symbol Lexer.KwForeign *> symbol Lexer.KwImport)
+    <*> foreignLangName
+    <*  symbol Lexer.BlockStart
+    <*> discover (nonEmptyBlock' foreignLocationImportList)
+
+foreignLocationImportList :: AsgParser SomeExpr
+foreignLocationImportList = buildAsg $
+    (\loc imports ->
+        liftAstApp2 IR.foreignLocationImpList' loc (sequence imports))
+    <$> stringOrVarName
+    <*  symbol Lexer.BlockStart
+    <*> discover (nonEmptyBlock' foreignSymbolImport)
+
+foreignSymbolImport :: AsgParser SomeExpr
+foreignSymbolImport = buildAsg $ withRecovery recover
+    $   try (foreignSymbolImportWithSafety defaultFISafety)
+    <|> foreignSymbolImportWithSafety specifiedFISafety
+    where
+        recover = (\e ->
+            invalid "Invalid safety specification."
+            <$ Loc.unregisteredDropSymbolsUntil
+            (`elem` [Lexer.ETX, Lexer.EOL]))
```

Comment:
We tend to use different style here - we try to have the "where" in the same line as variables, like:

```
...
where recover = ... 
```